Fixing temporarily from_pretrained_keras function

ZhanPascal opened the pull request:
To load the model, the function 'from_pretrained_keras' is using 'tf.keras.models.load_model', but it's providing a folder instead of the path to the model file So, we'll search for the first file with the .keras extension in the folder. If None is found then it will raise an error. If there is more than 1 model that ends .keras then raise error "Too many model".

Wauplin commented:
Hi @ZhanPascal, thanks for reporting and opening this PR :) Could you explain a bit more what is the problem you faced and which version of huggingface_hub/keras are you using? I'm just trying to figure out why this change is needed now.

ZhanPascal commented:
Hi @Wauplin 

To provide some context, I've used the following space on Hugging Face: [DracolIA_GPT2](https://huggingface.co/spaces/DracolIA/DracolIA_GPT2). The library versions I'm currently using are: Keras version 3.0.5 (latest version), TensorFlow version 2.15.0, and keras_nlp version 0.8.0.

I deployed the model on Hugging Face at this address: [DracolIA/GPT-2-LoRA-HealthCare](https://huggingface.co/DracolIA/GPT-2-LoRA-HealthCare), specifying it as belonging to the Keras library.

The code in question is:

```python
from huggingface_hub import from_pretrained_keras

model = from_pretrained_keras("DracolIA/GPT-2-LoRA-HealthCare")
```

The issue I encountered is with the `from_pretrained_keras` function, which calls `tf.keras.models.load_model`. This expects a `file_path` as an argument. However, `from_pretrained_keras` provides a `folder_path` (which is refered as `storage_folder` here). This behavior causes a mismatch as `load_model` cannot correctly handle a folder path.

I got this error :
```python
ValueError: File format not supported: filepath=/home/user/.cache/huggingface/hub/models--DracolIA--GPT-2-LoRA-HealthCare/snapshots/87ed572ea0e1cfc2c46c4ca5489109de716478a0.
```

To address this issue for my specific case, I modified the code to search within the specified folder for a model file ending with the `.keras` extension. I also added a check to ensure there are not multiple model files within the folder, as this could create ambiguity over which model to load.

I hope this explanation clarifies the issue encountered and the reason for the proposed modification in the Pull Request.

Best regards,

ZHAN Pascal

Wauplin commented:
Hi @ZhanPascal thanks a lot for the detailed explanation! The Keras x `huggingface_hub` integration is indeed broken by the Keras 2 -> 3 migration. This will become more prevalent now that latest tensorflow release makes keras 3 the default version.  To be honest the current integration is quite legacy but we should aim at keeping Keras2 compatibility. I just opened https://github.com/huggingface/huggingface_hub/pull/2107 to fail early when Keras3 is installed (instead of getting an unexpected error when executing the load_keras_pretrained and save_keras_pretrained helpers). The mid term objective is of course to support Keras3 models but I would prefer to think it through carefully and advocate to have it maintained directly in Keras repo - instead of here. In this regard, I hope it is fine for you if I put this PR on pause for now.

Wauplin commented:
Closing this PR without merging in favor of https://github.com/huggingface/huggingface_hub/pull/2107. As said in https://github.com/huggingface/huggingface_hub/pull/2104#issuecomment-1988934376 we favor a direct integration in Keras3 library rather than a semi-integration in `huggingface_hub`.
